packages/wrangler/src/publish.ts

```diff
@@ -36,10 +37,163 @@ type Props = {
   dryRun: boolean | undefined;
 };
 
+type Route = Exclude<Config["routes"], undefined>[number];
+type RouteObject<R extends Route = Route> = R extends string ? never : R;
+
 function sleep(ms: number) {
   return new Promise((resolve) => setTimeout(resolve, ms));
 }
 
+function renderRoute(route: Route): string {
+  let result = "";
+  if (typeof route === "string") {
+    result = route;
+  } else {
+    result = route.pattern;
+    const isCustomDomain = !!("custom_domain" in route && route.custom_domain);
+    if (isCustomDomain && "zone_id" in route) {
+      result += ` (custom domain - zone id: ${route.zone_id})`;
+    } else if (isCustomDomain && "zone_name" in route) {
+      result += ` (custom domain - zone name: ${route.zone_name})`;
+    } else if (isCustomDomain) {
+      result += ` (custom domain)`;
+    } else if ("zone_id" in route) {
+      result += ` (zone id: ${route.zone_id})`;
+    } else if ("zone_name" in route) {
+      result += ` (zone name: ${route.zone_name})`;
+    }
+  }
+  return result;
+}
+
+// this function takes a string with quotes in it
+// (i.e. `hello "world", if that really is your name`)
+// and peels out the first instance of a substring
+// bounded by quotes (so, in the example above, `world`)
+//
+// this is useful because the /domains api will return
+// which domains conflicted in an error message, bounded
+// by a string, which we can use to provide helpful
+// messages to a user
+function getQuoteBoundedSubstring(content: string) {
+  const matches = content.split('"');
+  return matches[1] ?? "";
+}
```

Comment:
Originally tried to use a regex here, but it was too permissive when there were multiple quote bounded substrings. This will always grab the first one (and even works when the string begins with a quote)

Reply:
there's no current situation where this could occur - I authored the code that generates these messages - but I suppose all bets are off if the code drifts. I could add a comment reinforcing the general structure in the api